prefetch-npm-deps: download dev deps for git deps with install scripts

winterqt opened the pull request:
###### Description of changes

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

winterqt committed:
prefetch-npm-deps: refactor

This splits prefetch-npm-deps into multiple files, as well as making a
few small changes along the way, such as going from a `HashMap` to a `Vec`
as the container for packages, to deduplicate them more efficently.

winterqt committed:
prefetch-npm-deps: download dev deps for git deps with install scripts

Git dependencies with install scripts are built isolated from the main
package, so their development dependencies are required.

To take advantage of this, #206477 is needed.

winterqt commented:
I wonder if I should combine this and #206477... hm.

lilyinstarlight commented on pkgs/build-support/node/fetch-npm-deps/src/parse/mod.rs:
```diff
@@ -0,0 +1,364 @@
+                    .ok()?,
+                )
+            }
+            _ => None,
```
We need a better error message here, because when a git URL is used that is not from a hosted git service, this `None` causes the calling code to assume it is a registry dep rather than git dep and fails saying some expected fields are missing (which are absent since it's not a registry dep)

winterqt replied:
The main issue with this is then deciding what's a registry dep and what's not. I fixed this by no longer checking if something is a Git dep if their scheme is `http` or `https`, as `hosted-git-info` does. (Do we even care about this case?)

lilyinstarlight replied:
Looks good now!

If you're asking whether we care to support arbitrary git URLs, I don't think so until someone brings it up as problem. I feel it's unlikely to ever come up, so there's no real reason to implement it until there is an actual need

winterqt replied:
> If you're asking whether we care to support arbitrary git URLs, I don't think so until someone brings it up as problem. I feel it's unlikely to ever come up, so there's no real reason to implement it until there is an actual need

It's impossible to do so with how npm implements it.

What I mean is if we should support URLs in the lockfile that look like `https://github.com/foo/bar`, when, as far as I know, they should always get resolved to `git+` protocols. (Look at the protocols we filter for in the old and new impl.)

lilyinstarlight replied:
Ah that's what you meant. Well they should always be resolved to `git+[proto]` so I don't think we need to worry about that

lilyinstarlight commented on pkgs/build-support/node/fetch-npm-deps/src/parse/mod.rs:
```diff
@@ -0,0 +1,364 @@
+            ),
+            (
+                "ssh://git@gitlab.com/foo/bar.git#fix/bug",
+                Some("https://gitlab.com/foo/bar/repository/archive.tar.gz?ref=fix/bug")
```
This URL is wrong but it's wrong in the upstream `hosted-git-info` package (see https://github.com/npm/hosted-git-info/pull/109) so idk if we really need to fix it here or not 🤷🏻‍♀️

lilyinstarlight commented on pkgs/build-support/node/fetch-npm-deps/default.nix:
```diff
@@ -36,11 +36,12 @@
-          src = makeTestSrc { inherit name src; };
-        };
+            src = makeTestSrc { inherit name src; };
+          } // lib.optionalAttrs forceGitDeps { I_LOVE_JS = true; });
```
Is this env var only for tests or is it possible it will be needed for actual packages?
If it might need to be used by actual packages, we should probably document it (maybe have it as a normal argument to the fetcher too?)

winterqt replied:
Fixed, except docs.

lilyinstarlight replied:
This one looks good now too except for a quick note in the documentation (or it could be in an error diagnostic message tbh)

winterqt replied:
Kind of against noting it in the documentation due to its ability to blow up things (maybe I'm being dumb), but I added it to `buildNpmPackage` + the error message in `prefetch-npm-deps`.

Also: I just realized that we may run into *Git deps with Yarn lockfiles* aaaaaaaaaaaa

lilyinstarlight replied:
Noting it just in the error message sounds great to me, and working on it after `buildYarnPackage` also sounds good to me

lilyinstarlight reviewed (commented):
I made a couple of nit comments, but overall the changes look good and function correctly in my testing

Edit: Also, the commit has `installl` with 3 `l`'s when it should just be `install`

winterqt commented:
Pushed the fixes as separate commits so you can more easily review them.

winterqt committed:
prefetch-npm-deps: throw better error when unsupported git service is used

See https://github.com/NixOS/nixpkgs/pull/206476#discussion_r1058689383.

winterqt committed:
prefetch-npm-deps: fix clippy lint

winterqt commented:
Okay, we should be good to go with this set of changes. Apologies for the delay as always.

@lilyinstarlight Please take a look, especially at the commit messages. Thanks.

lilyinstarlight reviewed (commented):
Okay so this looks good to me, but it seems to have broken the `mongosh` FOD

I haven't dived into why yet but it may just be that it needs to be updated because it wasn't generating right to begin with (and I guess it maybe worked by chance since it's built from a workspaced repo and doesn't need all locked deps idk). Regardless I'll investigate hopefully tomorrow so that this can get merged

I manually checked all other FODs and nixpkgs-review came in with no new regressions over master

Edit: The mongosh FOD hash changed becuase it wasn't downloading dev dependencies from https://github.com/mongodb-js/saslprep/tree/v1.0.4 (a transitive dependency) and so since including git dep lockfiles is the purpose of this PR, I will say that is Working As Intended and update the hash

lilyinstarlight committed:
mongosh: fix hash now that all git deps are fetched

This FOD hash changed becuase it the fetcher was not downloading dev
dependencies from https://github.com/mongodb-js/saslprep/tree/v1.0.4 (a
transitive dependency) and so since including git dep lockfiles is the
purpose of these changes this change is intended and the hash should be
updated.

Generally a package was unable to build without these so I did not
expect this to be a breaking change (since there would be no existing
FODs), but mongosh seems to have been an exception. I think just
changing the hash here is more appropriate for this exception rather
than working in machinery for supporting old and new FODs (since it's
rather unlikely these prefetch-npm-deps changes will break downstream
code as it should not have worked in the first place if it needed this).

lilyinstarlight commented:
I've done one push to rebase and another push to update the mongosh FOD hash (see commit message for why)

I think this should be good now, but I'll do more tests and solicit another review so I'm not self-reviewing

winterqt commented:
@ofborg eval

RaitoBezarius commented on pkgs/build-support/node/fetch-npm-deps/src/parse/mod.rs:
```diff
@@ -0,0 +1,370 @@
+                )
+            }
+            "gitlab.com" => {
+                /* let path = &url.path()[1..];
```
Is this code supposed to stay commented?

lilyinstarlight replied:
Yep upstream `hosted-git-info` npm package still does not work with gitlab.com so this URL just straight up does not work 🙃

lilyinstarlight replied:
(this is what the "lmao" on line 270 is referencing)

RaitoBezarius reviewed (approved):
This is awesome. :)

lilyinstarlight reviewed (approved):
I did one final pass and everything looks good, with no regressions against master (4 failing builds are also all failing in Hydra)

Merging! 🎉 🚀 

---

Result of `nixpkgs-review pr 206476` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>audiobookshelf</li> (<a href="https://hydra.nixos.org/job/nixpkgs/trunk/audiobookshelf.x86_64-linux">hydra failure</a>)
    <li>deltachat-desktop</li> (<a href="https://hydra.nixos.org/job/nixpkgs/trunk/deltachat-desktop.x86_64-linux">hydra failure</a>)
    <li>faustStk</li> (<a href="https://hydra.nixos.org/job/nixpkgs/trunk/faustStk.x86_64-linux">hydra failure</a>)
    <li>magnetophonDSP.ConstantDetuneChorus</li> (<a href="https://hydra.nixos.org/job/nixpkgs/trunk/magnetophonDSP.ConstantDetuneChorus.x86_64-linux">hydra failure</a>)
  </ul>
</details>
<details>
  <summary>69 packages built:</summary>
  <ul>
    <li>ansible-language-server</li>
    <li>authelia</li>
    <li>bitwarden</li>
    <li>cdxgen</li>
    <li>element-desktop</li>
    <li>element-desktop-wayland</li>
    <li>emscripten</li>
    <li>evcc</li>
    <li>faust (faust2)</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>guitarix</li>
    <li>hred</li>
    <li>iosevka</li>
    <li>iosevka-comfy.comfy</li>
    <li>iosevka-comfy.comfy-duo</li>
    <li>iosevka-comfy.comfy-fixed</li>
    <li>iosevka-comfy.comfy-motion</li>
    <li>iosevka-comfy.comfy-motion-duo</li>
    <li>iosevka-comfy.comfy-motion-fixed</li>
    <li>iosevka-comfy.comfy-wide</li>
    <li>iosevka-comfy.comfy-wide-duo</li>
    <li>iosevka-comfy.comfy-wide-fixed</li>
    <li>kapitonov-plugins-pack</li>
    <li>kaufkauflist</li>
    <li>lv_img_conv</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>matrix-appservice-irc</li>
    <li>mongosh</li>
    <li>mooSpace</li>
    <li>mpvScripts.webtorrent-mpv-hook</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>open-stage-control</li>
    <li>paperless-ngx</li>
    <li>photoprism</li>
    <li>pnpm-lock-export</li>
    <li>prefetch-npm-deps</li>
    <li>schildichat-desktop</li>
    <li>schildichat-desktop-wayland</li>
    <li>sharing</li>
    <li>sunshine</li>
    <li>tambura</li>
    <li>terminal-stocks</li>
    <li>torq</li>
    <li>twspace-crawler</li>
    <li>uptime-kuma</li>
    <li>vieb</li>
    <li>vsce</li>
    <li>webcord</li>
    <li>zigbee2mqtt</li>
    <li>zinc</li>
    <li>zrythm</li>
  </ul>
</details>
